notebooks/_06-Architecture.md

```diff
@@ -0,0 +1,633 @@
+
+
+
+
+
+### 13.4 What are the conventions for error messages that the system reports?
+
+
+
+
+
+### 13.5 Where are errors processed? At the point of detection, by a central error handling class, by functions above in the call stack, etc.
+
+
+
+
+
+### 13.6 What level of responsibility does each class have in validating its own input? Is there a central class (or set of classes) that performs all validation? When can classes assume clean information?
+
+
+
+
+
+### 13.7 How will we use (or not use) type hinting to help prevent errors?
+
+
+
+
+
+## 14. Description of testing plan:
+
+
+### 14.1 How will we test that all of the parts of the architecture are working correctly?
+- **How will we create tests for each part of the architecture that are:**
+ - **clear and unambiguous**
+ - **capable of dealing with stochastic functions / objects**
+ - **minimizing reliance / use of stochasticity**
+ - **fast to evaluate**
+
+
+### 14.2 What are the various types of tests that are needed?
+
+
+
+
+
+### 14.3 Are there any parts of the architecture that we do not know how to test? How will these knowledge gaps be resolved before the coding begins?
+
+
+
+
+
+### 14.4 How will discovered bugs be turned into test cases?
+
+
+
+
+
+### 14.5 How will the software be reviewed in general (beyond automated tests) to reduce the probability of programming bugs into the code / increase the chance of producing correctly functioning software?
+
+
+
+
+
+## 15. Description of project-sharing procedures:
+
+
+### 15.1 Is there a plan in place for producing documentation for the software?
+- **Documentation for operating the software**
+- **Documentation of software limits**
+- **Usage examples and tutorials**
+- **Key project learnings / retrospective notes**
+
+Documentation, in the form of Jupyter Notebook accompanied by a Markdown file (README.md) detailing the procedure for installing the environment and the content of folders and notebooks. Any limitations to the project will be noted in the jupyter notebooks developed to illustrate the analysis effort. The README.md markdown file will also will point to documentation pages for any packages used in the project.
+
+
+### 15.2 Is there a plan for sharing the software with others?
+- **File upload?**
+- **Upload to package repository?**
+
+All project related material including code, documentation, presentation, environment files, and datasets will be uploaded to the project github repository.
+
+
+### 15.3 Will the project have an associated website? If so is there a plan for its development?
+
+The only "website" for the project will be its github repository.
+
+
+### 15.4 Is there a plan for producing any written or visual materials that others may wish to view?
+- **Final or interim project reports**
+- **Tables of key project results**
+- **Graphs, diagrams, images from the project, e.g.:**
+ - **Plots of system performance**
+ - **Model checking plots**
+
```

Comment:
@timothyb0912 As far as I know, the only thing we will be producing will be a presentation to be sent to the conference organizers and uploaded to the repository. Maybe I am overthinking this question, but does it just ask for if we will be producing any presentable visual materials?

Reply:
Not if, but whether there is (and essentially what is) the plan for producing the visual materials?

What specific plots, tables, diagrams, etc. are we expecting to include?

What parts (e.g. what files / classes / functions) of the architecture / software are going to produce those materials?

Reply:
Got it.